akka-cluster/src/main/scala/akka/cluster/ClusterHeartbeat.scala

```diff
@@ -168,6 +177,261 @@ private[cluster] final class ClusterHeartbeatSender extends Actor with ActorLogg
+   *     `number-of-cross-datacenter-monitoring-actors`), so it does not have to monitor that other data centers
+   *
+   * In this state it will however listen to cluster events to eventually take over monitoring other DCs
+   * in case it becomes "old enough".
+   */
+  def dormant: Actor.Receive = {
+    case s: CurrentClusterState ⇒ init(s)
+    case MemberRemoved(m, _)    ⇒ removeMember(m)
+    case evt: MemberEvent       ⇒ addMember(evt.member)
+    case HeartbeatTick          ⇒ // ignore...
+  }
+
+  def active: Actor.Receive = {
+    case HeartbeatTick                ⇒ heartbeat()
+    case HeartbeatRsp(from)           ⇒ heartbeatRsp(from)
+    case MemberRemoved(m, _)          ⇒ removeMember(m)
+    case evt: MemberEvent             ⇒ addMember(evt.member)
+    case UnreachableMember(m)         ⇒ unreachableMember(m)
+    case ReachableMember(m)           ⇒ reachableMember(m)
+    case ExpectedFirstHeartbeat(from) ⇒ triggerFirstHeartbeat(from)
+  }
+
+  def introspecting: Actor.Receive = {
+    case ReportStatusTo(o) ⇒
+      log.debug(s"$incLocalClockJson Updated status observer to {} (activelyMonitoring:{})", o, activelyMonitoring)
+      if (activelyMonitoring) o.get ! CrossDcHeartbeatSender.MonitoringActive(state)
+      else o.get ! CrossDcHeartbeatSender.MonitoringDormant()
+      this.observer = o
+  }
+
+  def init(snapshot: CurrentClusterState): Unit = {
+    val nodes = snapshot.members.filter(filterExternalMembers).map(_.uniqueAddress)
+    val unreachable = snapshot.unreachable.filter(filterExternalMembers).map(_.uniqueAddress)
+    state = state.init(nodes, unreachable)
+  }
+
+  def addMember(m: Member): Unit = {
+    val memberIsCandidateForMonitoring =
+      m.status == MemberStatus.Up && // only Up members are considered for addition, since we require the oldest relationships to hold 
+        m.uniqueAddress != selfUniqueAddress && // is not self 
+        !state.contains(m.uniqueAddress) && // not already monitored
+        filterExternalMembers(m) // is outside of our datacenter
+
+    if (memberIsCandidateForMonitoring) {
+      log.debug(s"$incLocalClockJson Discovered candidate for monitoring ${m.address.port.get}, checking if old enough...")
+      log.debug(s"$incLocalClockJson Alpha members: ${cluster.membersByAge().filter(_.datacenter == "alpha").map(m ⇒ m.address.port.get + "@" + m.status)}")
+      log.debug(s"$incLocalClockJson Beta members: ${cluster.membersByAge().filter(_.datacenter == "beta").map(m ⇒ m.address.port.get + "@" + m.status)}")
+
+      // We only monitor the top N oldest members in each datacenter, those are considered the most "stable"
+      def shouldBeMonitored(m: Member): Boolean = {
+        // the number of nodes we monitor is the same as we have monitoring nodes:
+        val monitorActors = crossDcSettings.NrOfMonitoringActors
+        val res = cluster.isAtLeastNthOldestMember(m, _.datacenter == m.datacenter, monitorActors)
```

Comment:
I think it would be better if this actor itself had something like `Map[DataCenter, SortedSet[Member](ageOrdering)]` containing all members per DCs, and then the selection for each DC would be like a take(5),
and when a member is added/removed it would be easy to compare the sets before and after to see if there are any adjustments.

The Map can also contain the own DC, which can be used to decide if this is responsible for sending heartbeats.